modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java

```diff
@@ -96,44 +102,60 @@ public IgniteBiTuple<ClusterNode, Long> enlist(TablePartitionId tablePartitionId
     /** {@inheritDoc} */
     @Override
     protected CompletableFuture<Void> finish(boolean commit) {
-        Map<ClusterNode, List<IgniteBiTuple<TablePartitionId, Long>>> groups = new LinkedHashMap<>();
+        Map<TablePartitionId, Long> enlistedGroups = new LinkedHashMap<>();
+        CompletableFuture<Void> finishFuture;
 
         if (!enlisted.isEmpty()) {
-            enlisted.forEach((groupId, groupMeta) ->
-                    groups.computeIfAbsent(groupMeta.get1(), clusterNode -> new ArrayList<>())
-                            .add(new IgniteBiTuple<>(groupId, groupMeta.get2())));
+            enlistedGroups = enlisted.entrySet().stream()
+                    .collect(Collectors.toMap(
+                            Entry::getKey,
+                            entry -> entry.getValue().get2()
+                    ));
 
             IgniteBiTuple<ClusterNode, Long> nodeAndTerm = enlisted.get(commitPart);
 
             ClusterNode recipientNode = nodeAndTerm.get1();
             Long term = nodeAndTerm.get2();
 
             LOG.debug("Finish [recipientNode={}, term={} commit={}, txId={}, groups={}",
-                    recipientNode, term, commit, id(), groups);
+                    recipientNode, term, commit, id(), enlistedGroups);
 
             assert recipientNode != null;
             assert term != null;
 
-            return txManager.finish(
+            finishFuture = txManager.finish(
                     observableTsTracker,
                     commitPart,
                     recipientNode,
                     term,
                     commit,
-                    groups,
+                    enlistedGroups,
                     id()
             );
         } else {
-            return txManager.finish(
+            finishFuture = txManager.finish(
                     observableTsTracker,
                     null,
                     null,
                     null,
                     commit,
-                    groups,
+                    enlistedGroups,
                     id()
             );
         }
+
+        return finishFuture.exceptionally(
+                e -> {
+                    if (e instanceof IgniteInternalException) {
+                        // Transaction is already rolled back. Just wrapping internal exception with public one.
+                        throw withCause(TransactionException::new, TX_PRIMARY_REPLICA_EXPIRED_ERR,
+                                "Primary replica has expired, transaction will be rolled back", e);
```

Comment:
Why are you shure about the reason, when we checked an exception type only?

Reply:
You're right, instead of 
`if (e instanceof IgniteInternalException)`
we should have
`if (e instanceof PrimaryReplicaExpiredException)`
My fault. Fixed.